### ISSUE TITLE
Start is getting slower

### ISSUE THREAD

**Issue by s0cks** [ATLauncher/ATLauncher#192]
I find the start could be faster, need to find whats causing such slow start times

**Comment by RyanTheAllmighty** [comment]
There are a few things I want to do including removing the check for the servers.json at the beginning.

Also I believe we do a bit of work in querying Mojangs API for instance to check for username changes (which doesn't work in the background) which can be moved to the background, but we need to add in a 'AccountsReady' event that gets fired which prevents launching instances or editing accounts until that event is done.

The reason it's not done in the background is because it could modify account data, so shoving that to the background and having that event fire would be one way to decrease startup time by at least 500-1000 ms

**Comment by RyanTheAllmighty** [comment]
Okay I've done those 2 things I mentioned up there, seems to be faster starting up now, at least for me

**Comment by flaw600** [comment]
@s0cks is it better for you?

**Comment by s0cks** [comment]
@RyanTheAllmighty @flaw600 We could create a system that loads based upon tasks and each task is a thread and if it has a dependency it could be sorted and ran most optimally using say akka or I can try to create a smaller version of akka if that would suffice. It should theoretically make start times 10x faster. Plus if the start process needed to be changed it would be 100x easier too. Thoughts?

**Comment by s0cks** [comment]
Also @flaw600 could you join a irc net or something so we can chat about this?

**Comment by flaw600** [comment]
Why isn’t each task a thread already? Code complexity? I suppose that’s a valid reason.

**Comment by s0cks** [comment]
@flaw600 Not sure but it should be divided up like I proposed to get optimal load times.

**Comment by flaw600** [comment]
I’m on EsperNet. Doing some work atm but somewhat available to chat. I’ll be even more available to chat in 2 hours.

**Comment by flaw600** [comment]
Theoretically I’m not opposed to the plan, but I feel like we have enough issues we have to get through before we can worry about load times specifically.

**Comment by RyanTheAllmighty** [comment]
Honestly for a large bulk change like that sounds @s0cks I'd hold off for a 3.4 release since it sounds like a lot of code change and added complexity, but honestly no idea

**Comment by s0cks** [comment]
@RyanTheAllmighty Well with the lib that Im testing it with it seems like a simple fix, just a couple changes :P

**Comment by RyanTheAllmighty** [comment]
Yes well feel free to submit a PR so I can check it before adding it to the 3.3.0.0 codebase as I don't wont to make things too overly complex as I'm not able to contribute much anymore and Java isn't my language anymore so the less WTFing I have to do to fix issues the better :)

**Comment by s0cks** [comment]
Awesome I'll fork 3.3.0.0 sometime this week (maybe tonight if I have the time) and work on that :)

**Comment by RyanTheAllmighty** [comment]
Some improvements have been made to this today, but there's still some noticeable slowdown on startup:

-----

https://github.com/ATLauncher/ATLauncher/blob/master/src/main/java/com/atlauncher/utils/javafinder/JavaFinder.java#L60

This `findJavas` method currently takes around 1.5 seconds on my machine. The main reason is that it reads into the Windows registry to find the installed Java paths. We also check common program files paths to find installed Java versions, so we could remove the registry checking, but this would not show the Java's which are installed in non standard places, might be a trade off that's worth taking.

Note that this startup time increase is only present on Windows, other platforms do not do this check.

-----

The other place where startup time is increased is when we check for updates with the server.

One way we could overcome this is if the user has older files, use them and show the launcher, but do the update in the background and then load in the updated files (packs mostly) and reload the UI.

I think this would be quite difficult simply due to the way the UI is currently loaded in (see next point below).

-----

The next place where there's some noticable delay is when the data is all loaded, and the UI is loaded. Since all the tabs for the launcher are loaded into memory from the start, it has to take that time to load hundreds of packs and instances into memory ready to display.

Moving the packs, instances and servers tabs to be paginated, would help with that as it would limit the need to show more than say 10 or 20 things at a time, rather than all of them.

Could also be some improvements to the UI to lazy load the tabs contents, so it only loads the packs tab when a user actually clicks on it.

**Comment by RyanTheAllmighty** [comment]
Tab load timings:

```
[29/07/2020 22:34:38 pm] newsTab took 50 ms
[29/07/2020 22:34:38 pm] vanillaPacksTab took 70 ms
[29/07/2020 22:34:38 pm] featuredPacksTab took 50 ms
[29/07/2020 22:34:39 pm] packsTab took 607 ms
[29/07/2020 22:34:39 pm] instancesTab took 25 ms
[29/07/2020 22:34:39 pm] serversTab took 10 ms
[29/07/2020 22:34:39 pm] accountsTab took 20 ms
[29/07/2020 22:34:39 pm] toolsTab took 17 ms
[29/07/2020 22:34:41 pm] settingsTab took 2214 ms
```

**Comment by RyanTheAllmighty** [comment]
TechnicLauncher may have solved this issue.

Looking at https://github.com/TechnicPack/LauncherV3/blob/cffe12349766d978abf70381165869d0a35752ad/src/main/java/net/technicpack/launchercore/launch/java/source/os/WinRegistryJavaSource.java they use command line to essentially run:

`reg query HKEY_LOCAL_MACHINE\Software\JavaSoft\ /f Home /t REG_SZ /s /reg:64`

Which is super fast. Could be a good solution for the Java version finding

**Comment by RyanTheAllmighty** [comment]
Success. This call on Windows is now taking around 500ms, which is also offloaded to a thread on startup, so the impact is now minimal.

Paginating/lazy loading the tabs is next in the speed improvements

**Comment by RyanTheAllmighty** [comment]
Paginating the packs tab is quick and simple, so it's the solution for now. Need to keep an eye on the metrics around it and see if people are using it or not

**Comment by RyanTheAllmighty** [comment]
Released with version 3.4.0.0
